### ISSUE TITLE
Profiling Whittle and LoRA model

### ISSUE THREAD

**Issue by rheasukthanker** [whittle-org/whittle#293]
After profiling a subnet in whittle and the same subnet converted to litgpt, the litgpt subnet is about x10 faster. While the exact cause of this needs thorough investigation, following are my main suspets:
1. Set sub network calls get_qkv_indices and get_proj_indices, which is slow (this slows down training and fine-tuning
2. How expensive is slicing weights like we do compared to simple forward pass though a dense network?
3. Are we moving cos, sin buffers between devices in the forward pass?

Suggestion for future PRs: Every PR that introduces changes to the base model, or lora model, reports the increase/decrease in latency/profiling speed to make sure we have everything documented in PRs

**Comment by eddiebergman** [comment]
> Suggestion for future PRs: Every PR that introduces changes to the base model, or lora model, reports the increase/decrease in latency/profiling speed to make sure we have everything documented in PRs

Having some profiling script, even if not perfect would help a lot with this!

https://github.com/automl/ConfigSpace/tree/main/scripts

**Comment by aaronkl** [comment]
@eddiebergman @rheasukthanker I am wondering if we should integrate that into out CI somehow? Maybe some profiling script that create a some statistics as an artifact?

**Comment by Neonkraft** [comment]
TLDR; Whittle models are not 10x slower. There's a one-time _warmup_ on the first forward pass through the Whittle model.

I profiled the forward passes of both models in _train_ mode. If you profile only a single forward pass through the model, it would indeed appear as if Whittle is 10x slower.

A single forward pass on LitGPT takes less than 10ms
![Image](https://github.com/user-attachments/assets/b1196f7e-3734-4bfe-87f1-f661917fd32c)

A single forward pass on Whittle's GPT model takes over 100 ms
![Image](https://github.com/user-attachments/assets/2c33e49c-8275-471c-9130-dd62e087cb8e)

But if you consider multiple forward passes, the problem becomes clearer. There is a _one-time warmup_ cost for the Whittle model which is absent in LitGPT models. Consider ten forward passes, for example.

LitGPT (~80 ms):
![Image](https://github.com/user-attachments/assets/6014390d-81c4-42ff-97c2-855457df7e00)

Whittle GPT (~180ms):
![Image](https://github.com/user-attachments/assets/9f924609-2afa-4db1-b434-9fd40564f8f8)

This is a fixed, one-time cost. As the number of forward passes increases, the difference becomes negligible.

LitGPT (100 passes, ~800ms)
![Image](https://github.com/user-attachments/assets/697525af-34a4-4361-bf64-ae8bf92eab21)

Whittle GPT (100 passes, ~900ms)
![Image](https://github.com/user-attachments/assets/3c7da46f-8dbc-4b78-af60-819c9bd4fff9)

Unless we observe a slowdown in the actual training of the models, we can safely ignore this.

**Comment by Neonkraft** [comment]
You can find a gist with the code I used to profile the models [here](
https://gist.github.com/Neonkraft/00396b29eff5638a751959620b9c6866).

**Comment by Neonkraft** [comment]
UPDATE:

Whittle models incur a one-time cost every time `model.set_sub_network(**config)` is called. This means that training a _fixed_ sub-network will not take longer, but pretraining the supernet using strategies that call `set_sub_network` in every step (i.e., any strategy other than the `standard` strategy) is _drastically_ affected.

For a simple comparison, I compared `RandomStrategy` with 5 samples per step with a modified `StandardStrategy`, which performs forward and backward pass 5 times per step. `RandomStrategy` took twice as long as `StandardStrategy`.

The good news is that if we can find a way to circumvent the warmup, we can significantly improve the speed of pre-training the supernets.
